collie/model/hybrid_matrix_factorization.py

```diff
@@ -349,7 +480,71 @@ def forward(self, users: torch.tensor, items: torch.tensor) -> torch.tensor:
                 + self.user_biases(users)
                 + self.item_biases(items)
             )
+        elif self.hparams.stage in ('user_metadata_only', 'all') and self.item_metadata is None:
+            # TODO: remove self.device and let lightning do it
+            user_metadata_output = self.user_metadata[users, :].to(self.device)
+            if self.user_metadata_layers is not None:
+                for metadata_nn_layer in self.user_metadata_layers:
+                    user_metadata_output = self.dropout(
+                        F.leaky_relu(
+                            metadata_nn_layer(user_metadata_output)
+                        )
+                    )
+
+            # TODO: make this matrix factorization instead of only a MLP
+            combined_output = torch.cat((self.user_embeddings(users),
+                                         self.item_embeddings(items),
+                                         user_metadata_output), 1)
+            for combined_nn_layer in self.combined_layers[:-1]:
+                combined_output = self.dropout(
+                    F.leaky_relu(
+                        combined_nn_layer(combined_output)
+                    )
+                )
+
+            pred_scores = (
+                self.combined_layers[-1](combined_output)
+                + self.user_biases(users)
+                + self.item_biases(items)
+            )
+        else:
+            # TODO: remove self.device and let lightning do it
+            item_metadata_output = self.item_metadata[items, :].to(self.device)
+            if self.item_metadata_layers is not None:
+                for metadata_nn_layer in self.item_metadata_layers:
+                    item_metadata_output = self.dropout(
+                        F.leaky_relu(
+                            metadata_nn_layer(item_metadata_output)
+                        )
+                    )
+
+            # TODO: remove self.device and let lightning do it
+            user_metadata_output = self.user_metadata[users, :].to(self.device)
+            if self.user_metadata_layers is not None:
+                for metadata_nn_layer in self.user_metadata_layers:
+                    user_metadata_output = self.dropout(
+                        F.leaky_relu(
+                            metadata_nn_layer(user_metadata_output)
+                        )
+                    )
```

Comment:
[Optional] I have the same comment here about reducing the amount of times we copy-and-paste code several times in a file when it could be made into a helper method. I used to do this a lot in the past, but when I needed to make a change to that codeblock and forgot to make it to all the times it appears in a file, it lead to some hidden, nasty bugs that have scarred me for life. 

I know this is asking for a non-trivial refactor, so I'm happy to help out with some of the coding on this PR if that's not something you feel like doing - I want to be as helpful as possible here!